Save truth information in CaloPoints and add a few var

gputnam opened the pull request:
Depends on https://github.com/SBNSoftware/sbnanaobj/pull/90.

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18320&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6343&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6343&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18320&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11174&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11174&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18320&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18319&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6346&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6346&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18319&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11177&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11177&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18319&builds=)

brucehoward-physics commented:
I don’t think I’m necessarily the best reviewer here… for one, I’m not really familiar with the SCE functions, nor am I expert on everything being done here. In lieu of a review. I offer a few thoughts/questions I had while giving it a read, but again I’m not sure I’m really the best person to give the ultimate go ahead… 

* These are some fairly substantive additions. First, thanks for them! Second, curious what level(s) of validation has been done? 

* I think this is grabbing all the IDEs which match to a given readout channel for the given true particle? How is this affected by particle trajectory itself, e.g. say the trajectory is fairly into (or out of) the plane such that you end up making a number of hits on the same wire from the same track? Maybe not a concern? 

* How will looping through trajectory points be affected by things like particles being rolled up and such [maybe namely as an example, do deltas get actually saved as separate particles or are they tacked on in some way to the muon?]. Again maybe this is not a concern especially if everything is separated, but figured worth asking. 

* Looking at the comment on line 190 (https://github.com/SBNSoftware/sbncode/blob/3df2f0399c2a8a7308157b76e2ae8f63c08f2487/sbncode/CAFMaker/FillTrue.cxx#L190): a doxygen page I was looking at claimed it was the ionization position. Is it a quirk of how SCE is added within the LArG4 step that this comment is true? 

* Line 232 (https://github.com/SBNSoftware/sbncode/blob/3df2f0399c2a8a7308157b76e2ae8f63c08f2487/sbncode/CAFMaker/FillTrue.cxx#L232) - I guess the 1e-4 is just meant to be some arbitrarily small number to check that it’s actually been assigned something?

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18387&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18386&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11241&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11241&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18386&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11240&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11240&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18387&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6403&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6403&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18386&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6404&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6404&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18387&builds=)

gputnam commented:
To @brucehoward-physics 's comments:

1) This code is mostly taken from the truth matching in the Calibration N-tupler. So the validation basically comes from the operation of that module. 

2) Yes -- I think you'd just need to be careful about what the truth matching is telling you in that case.

3) Deltas would get rolled up into the main particle. I don't think though that the trajectory would usually be long enough for it to mess up the matching to the main particle trajectory.

4) The IDE's are after SCE

5) Yes i think so

Thanks for taking a look!

gputnam commented:
@miquelnebot it looks like the build failure is due to a versioning issue not related to this PR.

gputnam commented:
I've done a little more development on this code since I made the PR, so I just pushed another update. 

Still -- this PR is ready to go and should build. It works locally. When I checked the build failures here it looked like an issue unrelated to this PR.

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18522&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11321&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11321&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18522&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6516&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6516&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18522&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18523&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11323&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11323&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18523&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6518&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6518&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18523&builds=)

jacoblarkin reviewed (approved):
Similar to Bruce I'm not sure I'm the best person to review the physics of these changes.

But the code looks good and I'll approve with the caveat that I'm not the right person to review the physics. If this was validated in the calibration ntuples maybe that doesn't matter, but that's not my call. It would be useful to get Gianluca's review as well.

gputnam commented:
@miquelnebot on the latest commit I forgot to push a corresponding commit to sbnanaobj -- that's my bad.

I just fixed this, it should build now.

gputnam commented:
trigger build https://github.com/SBNSoftware/sbnanaobj/pull/90.

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18595&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18594&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11358&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11358&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18595&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6578&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6578&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18595&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6577&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6577&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18594&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11359&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11359&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18594&builds=)

gputnam commented:
trigger build https://github.com/SBNSoftware/sbnanaobj/pull/90

gputnam commented:
@miquelnebot when I check the build errors it doesn't seemed to have picked up the sbnanaobj PR -- did I do something wrong?

I also found an error in `mrbsetenv` in the build:

```125: ----------- check this block for errors -----------------------
126: Error encountered when setting up product: lardataobj
127: ERROR: Version conflict -- dependency tree requires versions conflicting with current setup of product lardataobj: version v09_13_04 vs v09_13_05
128: ERROR: setup -B lardataobj v09_13_04 -q c7:+prof failed
129: ERROR: setup of required products has failed```

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+        // residual range
+        truep.rr = 0;
+        if (traj_index >= 0) {
+          for (int i_traj = traj_index+1; i_traj < (int)particle.NumberTrajectoryPoints(); i_traj++) {
+            truep.rr += (particle.Position(i_traj).Vect() - particle.Position(i_traj-1).Vect()).Mag();
+          }
```
It may be guessed from my comments that I am a bit concerned of the time it takes to run this.
Here we are computing a residual range for each point of the track on each of the planes.
One thing that can be done to speed up _this_ part is to pre-compute the range as function of the trajectory point.
Something like:
```cpp
std::vector<double> getPartialRange(simb::MCTrajectory const& track) {
  if (track.empty()) return {};
  std::vector<double> ranges(track.size());
  std::size_t iPoint = track.size();
  ranges[--iPoint] = 0.0;
  while(iPoint-- > 0) {
    ranges[iPoint] = ranges[iPoint + 1] + (track.Position(iPoint+1) - track.Position(iPoint)).Vect().Mag();
  }
  return ranges;
} // getPartialRange()
```
should do (then `truep.rr = partialRanges[traj_index];`).

gputnam replied:
Fixed -- thanks for this!

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+        // directly apply the true direction. We include the effect of space charge
+        // on the true pitch here
+        if (closest_dist >= 0. && direction.Mag() > 1e-4) {
+          geo::PlaneID plane(srtrack.producer, p.tpc, iplane);
```
Please move this at the beginning of the plane loop, but like the full object:
```cpp
geo::PlaneGeo const& plane = geo->Plane({ srtrack.producer, p.tpc, iplane });
```
then we can avoid all that nonsense about views.

gputnam replied:
This can't be moved to the Plane loop because the TPC depends on the particle position.

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+        // on the true pitch here
+        if (closest_dist >= 0. && direction.Mag() > 1e-4) {
+          geo::PlaneID plane(srtrack.producer, p.tpc, iplane);
+          float angletovert = geo->WireAngleToVertical(geo->View(plane), plane) - 0.5*::util::pi<>();
+          TVector3 loc_mdx_v = loc_nosce_v - direction * (geo->WirePitch(geo->View(plane)) / 2.);
+          TVector3 loc_pdx_v = loc_nosce_v + direction * (geo->WirePitch(geo->View(plane)) / 2.);
+
+          // Convert types for helper functions
+          geo::Point_t loc_mdx(loc_mdx_v.X(), loc_mdx_v.Y(), loc_mdx_v.Z());
+          geo::Point_t loc_pdx(loc_pdx_v.X(), loc_pdx_v.Y(), loc_pdx_v.Z());
```
Assuming `direction` to be defined as `geo::Vector_t` instead of `TVector3` (you can use a `geo::vect::toVector()` on line 239, for example):
```suggestion
          geo::Point_t loc_mdx = loc_nosce - direction * plane.WirePitch() / 2.;
          geo::Point_t loc_pdx = loc_nosce + direction * plane.WirePitch() / 2.;
```
I pruned `angletovert` (it would have been `double const angletovert = plane.PhiZ();`) because it won't be needed, see below.

gputnam replied:
This code copies the interfaces used in the Calorimetry code for calculating pitch. I'd rather keep the use of interfaces consistent. I've been burned before by code behavior changing in LArSoft services and I'd rather reduce the number of different interfaces I use overall.

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+
+          // Map to wires
+          if (sce && sce->EnableSimSpatialSCE()) { 
+            int corr = geo->TPC(plane.TPC).DriftDir()[0];
```
```suggestion
            int corr = -plane.GetNormalDirection().X();
```
(note that the normal of the plane points to the centre of the TPC, so it's opposite to the drift direction)

gputnam replied:
See above comment.

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+          }
+
+          // Direction at wires
+          geo::Vector_t dir = (loc_pdx - loc_mdx) /  (loc_mdx - loc_pdx).r();
```
```suggestion
          geo::Vector_t dir = (loc_pdx - loc_mdx).unit();
```
I recommend a name that does not get confused with the other variable, `direction`.

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+          double cosgamma = std::abs(std::sin(angletovert)*dir.Y() + std::cos(angletovert)*dir.Z());
+          double pitch;
+          if (cosgamma) {
+            pitch = geo->WirePitch(geo->View(plane))/cosgamma;
```
```suggestion
            pitch = plane.WirePitch()/cosgamma;
```

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+          geo::Vector_t dir = (loc_pdx - loc_mdx) /  (loc_mdx - loc_pdx).r();
+
+          // Pitch at wires
+          double cosgamma = std::abs(std::sin(angletovert)*dir.Y() + std::cos(angletovert)*dir.Z());
```
```suggestion
          double cosgamma = std::abs(plane.SinPhiZ()*dir.Y() + plane.CosPhiZ()*dir.Z());
```
Short explanation: `geo::WireAngleToVertical()` [is actually not the wire angle to vertical](https://cdcvs.fnal.gov/redmine/issues/27790), but it is the wire angle (equivalent to `geo::PlaneGeo::ThetaZ()`), and that is the reason while in the definition a few lines above in your code its value was rotated by a right angle. `geo::PlaneGeo` also provides the direction orthogonal to the wires (i.e. the direction that the wire plane measures), calling it `PhiZ()`, and since _z_ is "the horizontal" and _y_ is "the vertical", `PhiZ()` happens to be also the angle _of the wire_ from the vertical (barred possibly a sign, which in `PhiZ()` is defined as the direction where the coordinate increases).
Finally, the plane provides pre-computed sine and cosine.

gputnam replied:
See above comment. I agree that the existing code is kind of confusing. But it gives the correct result and is the same interface that the Calorimetry uses.

PetrilloAtWork commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+          geo::Point_t loc_atwires_alongpitch = loc_sce + dir*pitch; 
+          geo::Point_t loc_alongpitch;
+          if (sce && sce->EnableSimSpatialSCE()) { 
+            geo::Vector_t offset = sce->GetCalPosOffsets(loc_atwires_alongpitch, plane.TPC);
```
With the changes above, this may become:
```suggestion
            geo::Vector_t offset = sce->GetCalPosOffsets(loc_atwires_alongpitch, p.TPC);
```
(or `plane.ID().TPC`)

PetrilloAtWork reviewed (commented):
My concern is how much computing time this adds.
I have recommended some changes that include:
* small optimisations to speed code up
* straighter use of the `Geometry`

Please take a look at them and consider whether to adopt them.
Also, please report on whether the CPU time increase is noticeable/negligible/dramatic.

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+        truep.y = 0.;
+        truep.z = 0.;
+
+        for (const sim::IDE *ide: ides) {
+          truep.x = (truep.x*truep.p_nelec + ide->x*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+          truep.y = (truep.y*truep.p_nelec + ide->y*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+          truep.z = (truep.z*truep.p_nelec + ide->z*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+
+          truep.p_e += ide->energy;
+          truep.p_nelec += ide->numElectrons;
+        }
```
I think the existing averaging code is fine. This doesn't change behavior and to me is a less intuitive way of computing an averge.

PetrilloAtWork replied:
It's also three times slower (counting multiplications).
I am a bit surprised you find that more intuitive, but I won't certainly argue about that.

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+    // Load the hits
+    // match on the channel, which is unique
+    const std::vector<std::pair<geo::WireID, const sim::IDE*>> match_ides = id_to_ide_map.at(srtrack.truth.p.G4ID);
+    std::map<unsigned, std::vector<const sim::IDE*>> chan_2_ides;
+    for (auto const &ide_pair: match_ides) {
+      chan_2_ides[geo->PlaneWireToChannel(ide_pair.first)].push_back(ide_pair.second);
+    }
```
ide_pair.second is a sim::IDE pointer, not a vector of pointers. So this wouldn't work.

PetrilloAtWork replied:
True, my bad (I).

gputnam commented:
I've addressed the comments above. Also, regarding performance concerns. I've added in a fcl parameter to save or not save the CaloTruth information, so it can be tuned by analyzers as necessary. 

Also, I've done a quick performance test on 15 events (nu+cosmic) with the latest updates:

Without saving truth information

```
TimeTracker printout (sec)                                       Min           Avg           Max         Median          RMS         nEvts
runprod:cafmaker:CAFMaker                                     0.364164       1.06376       2.15046       1.08088      0.408604        15
```

With Saving truth information:

```
TimeTracker printout (sec)                                       Min           Avg           Max         Median          RMS         nEvts
runprod:cafmaker:CAFMaker                                     0.710364       2.18523       4.16181       2.1704       0.834769        15
```

So, it doubles the time for CAFMaker. This is significant, but it's still only 2 seconds and is smaller than other processes run at the same stage (mcreco).

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+    const simb::MCParticle &particle = *match;
+
+    // Get service
+    auto const* sce = lar::providerFrom<spacecharge::SpaceChargeService>();
```
Fixed.

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+
+    // Load the hits
+    // match on the channel, which is unique
+    const std::vector<std::pair<geo::WireID, const sim::IDE*>> match_ides = id_to_ide_map.at(srtrack.truth.p.G4ID);
```
Fixed.

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+        // std::vector<sim::TrackIDE> ides = bt_serv->HitToTrackIDEs(clockData, hit);
+        //
+        // Use this:
+        art::ServiceHandle<cheat::BackTrackerService> bt_serv;
```
Fixed.

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+          continue; 
+        }
+
+        const std::vector<const sim::IDE *> &ides = chan_2_ides.at(p.channel);
```
Can't do this

gputnam commented on sbncode/CAFMaker/FillTrue.cxx:
```diff
@@ -127,6 +130,190 @@ namespace caf {
+            closest_dist = (particle.Position(i_traj).Vect() - loc_nosce_v).Mag();
+            traj_index = i_traj;
+          }
+        }
```
I'd rather use the TVector's -- also see the performance results below.

PetrilloAtWork reviewed (commented):
I have left a few follow up comments, but it's up to you now.
I appreciate the addition of the option not to add this truth information.
Whatever you decide, I would be interested in trying to profile this code after it's merged — so if you can paste the command line you used to test it, it will make profiling easier for me. While in absolute terms it's indeed invisible in comparison with other stages, if there is an easy way to halve the execution time, I would go for it (that's what I heard being called a "low-hanging fruit").

gputnam commented:
@PetrilloAtWork I fixed the Mag2 thing. 

On the interface comment, I appreciate the issue you are raising. I agree that the code you suggest would likely be an improvement. However, I don't think that the performance issues are a bottleneck and I have worries about the stability of the interface that you are proposing to use. I think that a possible issue to raise would be to switch over all of the relevant code in sbncode to using the improved interfaces for calculating the pitch. I would want that to be done all at once if we decided to.

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18636&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18637&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11389&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11389&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18637&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11388&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11388&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18636&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6605&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6605&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18637&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6606&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6606&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18636&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18634&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18635&builds=)

PetrilloAtWork reviewed (approved):
Official Chrism of Approval.

FNALbuild commented:
:x: CI build for SBND Failed at phase setup_environment_cvmfs SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11398&builds=)

:rotating_light: For more details about the failed phase, check the [setup_environment_cvmfs SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11398&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=setup_environment_cvmfs&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18634&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase setup_environment_cvmfs SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11399&builds=)

:rotating_light: For more details about the failed phase, check the [setup_environment_cvmfs SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11399&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=setup_environment_cvmfs&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18635&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase setup_environment_cvmfs ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6617&builds=)

:rotating_light: For more details about the failed phase, check the [setup_environment_cvmfs ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6617&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=setup_environment_cvmfs&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18634&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase setup_environment_cvmfs ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6618&builds=)

:rotating_light: For more details about the failed phase, check the [setup_environment_cvmfs ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6618&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=setup_environment_cvmfs&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18635&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18653&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18654&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6624&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6624&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18653&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11401&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11401&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18653&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11400&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11400&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18654&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6625&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6625&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18654&builds=)

gputnam commented:
Ok pushed an update that should fix the remaining build issues.

gputnam commented:
trigger build https://github.com/SBNSoftware/sbnanaobj/pull/90

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18671&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18672&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6640&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6640&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18671&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11416&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11416&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18671&builds=)

FNALbuild commented:
:x: CI build for ICARUS Failed at phase build ICARUS on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6641&builds=)

:rotating_light: For more details about the failed phase, check the [build ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6641&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18672&builds=)

FNALbuild commented:
:x: CI build for SBND Failed at phase build SBND on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11417&builds=)

:rotating_light: For more details about the failed phase, check the [build SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11417&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=build&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18672&builds=)

gputnam commented:
oh looks like I did the build command wrong

gputnam commented:
trigger build SBNSoftware/sbnanaobj#90

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18683&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/18684&builds=)

FNALbuild commented:
:warning: CI build for SBND Warning at phase ci_tests SBND on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/11420&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests SBND phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/11420&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18684&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for c7:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6650&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6650&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18683&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/6651&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/6651&platform=Linux%203.10.0-1160.81.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/18684&builds=)
